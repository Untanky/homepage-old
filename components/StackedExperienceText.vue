<template lang="pug">
  div
    .text-xl.font-bold {{element.title}}
    .italic {{element.subtitle}}
    .text-gray-600 {{element.description}}
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'StackedExperienceText',
  props: {
    element: {
      type: Object,
      required: true,
    },
  },
});
</script>
