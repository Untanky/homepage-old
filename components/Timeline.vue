<template lang="pug">
  .timeline
    .relative
      div(v-for="(element, index) in elements.list", :key="index")
        timeline-event(:element="element.timelineEntry")
          slot(:name="`timeline-event__${index}`")
</template>

<script lang="ts">
import Vue, { PropType } from 'vue';
import TimelineEvent from './TimelineEvent.vue';
import TimelineList from '../src/TimelineList';

export default Vue.extend({
  name: 'Timeline',
  components: {
    TimelineEvent,
  },
  props: {
    elements: {
      type: Object as PropType<TimelineList>,
      required: true,
    },
  },
});
</script>

<style lang="postcss" scoped>
.timeline {
  @apply p-4;
  @apply h-auto;
  @apply relative;
}

.timeline > .relative::before {
  content: "";
  @apply w-1;
  @apply h-full;
  @apply block;
  @apply absolute;
  @apply bg-gray-600;
  @apply left-7;
}

.timeline > .relative > *:first-child::before {
  content: "";
  @apply block;
  @apply absolute;
  @apply bg-white;
  @apply w-1;
  @apply h-4;
  @apply left-7;
}

.timeline > .relative > *:last-child {
  @apply relative;
}

.timeline > .relative > *:last-child > *::before {
  content: "";
  @apply block;
  @apply absolute;
  @apply bg-white;
  @apply w-1;
  @apply h-full;
  @apply left-7;
  @apply -bottom-4;
}
</style>
