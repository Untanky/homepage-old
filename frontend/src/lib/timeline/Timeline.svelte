<script lang="ts">
  import type { SvelteComponentTyped } from "svelte";
  
  export let elements: any[];
  export let component: typeof SvelteComponentTyped<{ element: any }>;
</script>

<ol class="relative">
  {#each elements as element }
    <li class="timeline-item">
      <div class="relative pb-6">
        <div class="line">
          <svelte:component this={component} element={element} />
        </div>
      </div>
    </li>
  {/each}
</ol>

<style>
  .timeline-item {
    @apply pl-8;
  }

  .timeline-item::before {
    content: "";
    @apply block absolute w-2 h-2 left-2 mt-2.5 rounded-full bg-sky-700 z-50;
  }

  .timeline-item:not(:last-child) .line::before {
    content: "";
    @apply block absolute w-1 h-full -left-[22px] top-3 rounded-full bg-sky-400;
  }
</style>
