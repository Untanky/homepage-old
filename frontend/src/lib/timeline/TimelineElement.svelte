<script lang="ts" context="module">
  export interface TimelineElement {
    header: string;
    formattedTime: string;
    description: string[];
  }
</script>

<script lang="ts">
  export let element: TimelineElement;
</script>

<h3 class="text-xl font-medium mb-1">{ element.header }</h3>
<i class="block text-xl italic mb-1">{ element.formattedTime }</i>
{#each element.description as paragraph}
  <p class="text-gray-600">{ paragraph }</p>
{/each}
