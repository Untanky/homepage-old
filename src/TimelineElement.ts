import TimelineEntry from './TimelineEntry';

class TimelineElement {
  content: any;

  timelineEntry: TimelineEntry;

  constructor(content: any, timelineEntry: TimelineEntry) {
    this.content = content;
    this.timelineEntry = timelineEntry;
  }
}

export default TimelineElement;
