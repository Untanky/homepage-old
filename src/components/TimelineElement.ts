import TimelineEntry from './TimelineEntry';

interface TimelineElement {
  content: any;
  timelineEntry: TimelineEntry;
}

export default TimelineElement;
