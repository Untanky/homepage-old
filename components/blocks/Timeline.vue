<template lang="pug">
  .timeline
    .relative
      div(v-for="(element, index) in elements.list", :key="index")
        timeline-event(:element="element.timelineEntry")
          slot(:name="`timeline-event__${index}`")
</template>

<script lang="ts">
import Vue, { PropType } from 'vue';
import TimelineEvent from './TimelineEvent.vue';
import { TimelineList } from '../../src/components';

export default Vue.extend({
  name: 'Timeline',
  components: {
    TimelineEvent,
  },
  props: {
    elements: {
      type: Object as PropType<TimelineList>,
      required: true,
    },
  },
});
</script>

<style lang="postcss" scoped>
.timeline {
  @apply h-auto;
  @apply relative;
  @apply p-4;
  @apply my-4;
}

.timeline > .relative::before {
  content: "";
  @apply w-1;
  @apply h-full;
  @apply block;
  @apply absolute;
  @apply bg-gray-600;
  @apply left-7;
}

.timeline > .relative > *:first-child::before {
  content: "";
  @apply block;
  @apply absolute;
  @apply bg-white;
  @apply w-1;
  @apply h-5;
  @apply left-7;
}

.timeline > .relative > *:last-child {
  @apply relative;
}

.timeline > .relative > *:last-child > *::before {
  content: "";
  @apply block;
  @apply absolute;
  @apply bg-white;
  @apply w-1;
  @apply h-full;
  @apply left-7;
  @apply -bottom-4;
}
</style>
