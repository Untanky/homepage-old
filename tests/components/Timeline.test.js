/* eslint-disable no-undef */
// global it, expect
import { mount } from '@vue/test-utils';
import { DateTime } from 'luxon';
import Timeline from '../../components/Timeline.vue';
import StackedExperienceText from '../../components/StackedExperienceText.vue';
import TimelineElement from '../../src/TimelineEntry';

it('Timeline should render correctly', () => {
  const props = {
    elements: [
      new TimelineElement(
        'Test',
        'Subtest',
        'Very long description',
        StackedExperienceText,
        DateTime.fromObject({ year: 2020, month: 11 }),
        DateTime.fromObject({ year: 2020, month: 11 }),
        true,
      ),
      new TimelineElement(
        'Test 2',
        'Subtest - even more test',
        'Very long description - much longer than before',
        StackedExperienceText,
        DateTime.fromObject({ year: 2020, month: 11 }),
        null,
        false,
      ),
      new TimelineElement(
        'Test 3',
        'Subtest - even more test',
        'Very long description - much longer than before',
        StackedExperienceText,
        DateTime.fromObject({ year: 2020, month: 11 }),
        null,
        true,
      ),
    ],
  };

  const comp = mount(Timeline, { propsData: props });
  expect(comp.element).toMatchSnapshot();
});
