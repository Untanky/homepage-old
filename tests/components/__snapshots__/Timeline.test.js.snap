// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline should render correctly 1`] = `
<div
  class="timeline"
>
  <div
    class="relative"
  >
    <div>
      <div
        class="timeline-event active"
      >
        <div
          class="text-xl font-medium"
        >
          November 2020 - November 2020
        </div>
        <div>
          <div
            class="text-xl font-bold"
          >
            Test
          </div>
          <div
            class="italic"
          >
            Subtest
          </div>
          <div
            class="text-gray-600"
          >
            Very long description
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="timeline-event"
      >
        <div
          class="text-xl font-medium"
        >
          November 2020
        </div>
        <div>
          <div
            class="text-xl font-bold"
          >
            Test 2
          </div>
          <div
            class="italic"
          >
            Subtest - even more test
          </div>
          <div
            class="text-gray-600"
          >
            Very long description - much longer than before
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="timeline-event active"
      >
        <div
          class="text-xl font-medium"
        >
          November 2020 - Now
        </div>
        <div>
          <div
            class="text-xl font-bold"
          >
            Test 3
          </div>
          <div
            class="italic"
          >
            Subtest - even more test
          </div>
          <div
            class="text-gray-600"
          >
            Very long description - much longer than before
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
